functions/AWSECSWrangleASG.js

```diff
@@ -0,0 +1,174 @@
+ * - If it is an EC2 backed auto scaling group it will:
+ *   - Evaluate all of the instances capacity in the ASG
+ *   - If it sees an opportunity to remove an EC2 instance from the ASG,
+ *   - (without changing the quantity and type of tasks running on the ASG)
+ *   - It will gracefully drain and terminate that instance
+ *   - (and downsize the ASG)
+ * 
+ * @param {String} OW_REGION Region
+ * @param {String} OW_CLUSTER Cluster name
+ * @param {String} [OW_CAPACITYPROVIDER] Capacity Provider
+ * 
+ */
+
+// These 2 imports are used to check if the function is run directly from the CLI vs as an ES6 module
+import { fileURLToPath } from 'url';
+import process from 'process';
+
+import {
+    ECSClient,
+    ListContainerInstancesCommand,
+    DescribeContainerInstancesCommand,
+    UpdateContainerInstancesStateCommand
+} from "@aws-sdk/client-ecs";
+
+import { 
+    AutoScalingClient, 
+    TerminateInstanceInAutoScalingGroupCommand
+} from "@aws-sdk/client-auto-scaling";
+
+const drainingInstanceTimeoutSeconds = 60;
+
+const reWrangleTimeoutSeconds = 10;
+
+export default async function run() {
+
+    try {
+        
+        const params = {
+            region: arguments[0] || process.env.OW_REGION || "",
+            cluster: arguments[1] || process.env.OW_CLUSTER || "",
+            capacityprovider: arguments[2] || process.env.OW_CAPACITYPROVIDER || ""
+        };
+
+        const ecsclient = new ECSClient(params);
+        const asgclient = new AutoScalingClient(params);
+        
+        // Get the active containerInstanceArns given the cluster name
+        const containerInstances = await ecsclient.send(
+            new ListContainerInstancesCommand({
+                cluster: params.cluster,
+                status: "ACTIVE",
+            }))
+        // Get the detail for those containerInstanceArns
+        const describeContainerInstances = await ecsclient.send(
+            new DescribeContainerInstancesCommand({
+                cluster: params.cluster,
+                containerInstances: containerInstances.containerInstanceArns
+            }))
+
+        // If capacityprovider is specified, parse the container list for only the containers in that capacity provider
+        let containersToEvaluate = []; 
+        if (params.capacityprovider) {
+            containersToEvaluate = describeContainerInstances.containerInstances.filter(record => record.capacityProviderName == params.capacityprovider);
+        }
+        else {
+            containersToEvaluate = describeContainerInstances.containerInstances;
+        }
+
+        console.log(`Running AWSECSWrangleASG on cluster: ${params.cluster} and Capacity Provider: ${containersToEvaluate[1].capacityProviderName}`);
+
+        // Calculate the container instance with the most available resources
+        // Calculate total available resources
+        // Calculate used resources per container instance 
+        let instanceWithMostResources = {
```

Comment:
Should be a `const` (same goes for any variable that doesn't need to be re-assigned)